cunumeric/array.py

```diff
@@ -958,12 +958,84 @@ def byteswap(self, inplace=False):
             numpy_array = self.__array__(stacklevel=3).byteswap(inplace=False)
             return self.convert_to_cunumeric_ndarray(numpy_array, stacklevel=3)
 
-    @unimplemented
-    def choose(self, choices, out, mode="raise"):
-        numpy_array = self.__array__(stacklevel=3).choose(
-            choices=choices, out=out, mode=mode
+    def choose(self, choices, out=None, mode="raise", stacklevel=2):
+        a = self
+        if out is not None:
+            out = out.convert_to_cunumeric_ndarray(out)
+
+        if isinstance(choices, list):
+            choices = tuple(choices)
+        is_tuple = isinstance(choices, tuple)
+        if is_tuple:
+            n = len(choices)
+            dtypes = [ch.dtype for ch in choices]
+            ch_dtype = np.find_common_type(dtypes, [])
+            choices = tuple(
+                self.convert_to_cunumeric_ndarray(choices[i]).astype(ch_dtype)
+                for i in range(n)
+            )
+
+        else:
+            choices = self.convert_to_cunumeric_ndarray(choices)
+            n = choices.shape[0]
+            ch_dtype = choices.dtype
+            choices = tuple(choices[i, ...] for i in range(n))
+
+        if not np.issubdtype(self.dtype, np.integer):
```

Comment:
This check is insufficient. On the C++ side you assume a specific datatype for the indices array when you're creating the accessor, and accessor creation needs to know the exact bit-width of the data elements. Therefore you need to ensure the type of the incoming data is exactly what the C++ code expects, not just a subtype. You can either generalize the C++ code to handle index arrays of various widths (that's likely overkill), require the incoming indices array to have the expected type (`np.int64`), or cast `a` to the correct type if necessary (creating a copy).